// Copyright (c) farm-ng, inc.
//
// Licensed under the Amiga Development Kit License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://github.com/farm-ng/amiga-dev-kit/blob/main/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Copyright (c) farm-ng, inc. All rights reserved.

syntax = "proto3";


import "farm_ng/core/linalg.proto";
import "farm_ng/core/pose.proto";

package farm_ng.filter.proto;

enum DivergenceReason {
  NO_DIVERGENCE = 0; // Indicates normal operation without divergence
  GPS_ERROR_X_HIGH = 1;
  GPS_ERROR_Y_HIGH = 2;
  STD_DEV_X_HIGH = 3;
  STD_DEV_Y_HIGH = 4;
  STD_DEV_W_HIGH = 5;
  NO_GYRO_MEASUREMENTS = 6;
  GYRO_MEASUREMENTS_DELAYED = 7;
  OTHER = 8; // For cases not covered above
}

message FilterState {
  // UKF state
  farm_ng.core.proto.Pose pose = 1;
  bool has_converged = 2;
  bool is_calibrated = 3;

  // Std deviation of the state, from the covariance matrix diagonal
  farm_ng.core.proto.VecXF64 uncertainty_diagonal = 4;
  farm_ng.core.proto.VecXF64 innovation = 5;

  double heading = 6; // Rotation about the z-axis contained in pose, in radians
  repeated DivergenceReason divergence_reasons = 7; // Reason for divergence, if any
}

message FilterTrack {
  // Deprecation warning: The FilterTrack message is deprecated.
  // Please use the generic Track message in track.proto instead.
  repeated FilterState states = 1;
  string name = 2;
}
